capsules/src/virtual_ninedof.rs

```diff
@@ -0,0 +1,141 @@
+//! This allows a 9DOF sensor to be used by multiple apps.
+//!
+//! The data from the driver is not virtualized. This just gives apps
+//! exclusive access to the driver until a callback occurs.
+
+use core::cell::Cell;
+use kernel::{AppId, Callback, Container, Driver};
+use kernel::hil;
+use kernel::returncode::ReturnCode;
+
+pub struct App {
+    callback: Option<Callback>,
+    pending_command: bool,
+    command: usize,
+    arg1: usize,
+}
+
+impl Default for App {
+    fn default() -> App {
+        App {
+            callback: None,
+            pending_command: false,
+            command: 0,
+            arg1: 0,
+        }
+    }
+}
+
+pub struct VirtualNineDof<'a> {
+    driver: &'a hil::ninedof::NineDof,
+    apps: Container<App>,
+    current_app: Cell<Option<AppId>>,
+}
+
+impl<'a> VirtualNineDof<'a> {
+    pub fn new(driver: &'a hil::ninedof::NineDof, container: Container<App>) -> VirtualNineDof<'a> {
+        VirtualNineDof {
+            driver: driver,
+            apps: container,
+            current_app: Cell::new(None),
+        }
+    }
+
+    fn enqueue_command(&self, command_num: usize, arg1: usize, appid: AppId) -> ReturnCode {
+        self.apps
+            .enter(appid, |app, _| {
+                // Check so see if we are doing something. If not,
+                // go ahead and do this command. If so, this is queued
+                // and will be run when the pending command completes.
+                if self.current_app.get().is_none() {
+                    self.current_app.set(Some(appid));
+                    self.call_driver(command_num, arg1)
+                } else {
+                    if app.pending_command == true {
+                        ReturnCode::ENOMEM
+                    } else {
+                        app.pending_command = true;
+                        app.command = command_num;
+                        app.arg1 = arg1;
+                        ReturnCode::SUCCESS
+                    }
+                }
+            })
+            .unwrap_or(ReturnCode::FAIL)
+    }
+
+    fn call_driver(&self, command_num: usize, _: usize) -> ReturnCode {
+        match command_num {
+            1 => self.driver.read_accelerometer(),
+            100 => self.driver.read_magnetometer(),
+            _ => ReturnCode::FAIL,
+        }
+    }
+}
+
+impl<'a> hil::ninedof::NineDofClient for VirtualNineDof<'a> {
+    fn callback(&self, arg1: usize, arg2: usize, arg3: usize) {
+        // Notify the current application that the command finished.
+        self.current_app.get().map(|appid| {
+            self.current_app.set(None);
+            let _ = self.apps.enter(appid, |app, _| {
+                app.pending_command = false;
+                app.callback.map(|mut cb| {
+                    cb.schedule(arg1, arg2, arg3);
+                });
+            });
```

Comment:
As an optimization, does it make sense to iterate over all other apps and schedule a callback for any app that has this command as pending? Of note, I think this only works because `arg1` from the `command` happens to be meaningless in the current implementation.